maraboupy/MarabouNetworkNNet.py

```diff
@@ -13,40 +13,180 @@
 MarabouNetworkNNet represents neural networks with piecewise linear constraints derived from the NNet format
 '''
 
-from .MarabouUtils import *
+from maraboupy.MarabouUtils import *
 from maraboupy import MarabouCore
 from maraboupy import MarabouNetwork
+
+import warnings
 import numpy as np
+import sys
+
 
 class MarabouNetworkNNet(MarabouNetwork.MarabouNetwork):
-    """Constructs a MarabouNetworkNNet object from an .nnet file
-
-    Args:
-        filename: path to the .nnet file
-        normalize: (bool) True if network parameters should be adjusted to incorporate
-                      network input/output normalization. Otherwise, properties must be written
-                      with the normalization already incorporated.
-    Attributes:
-        numLayers (int): The number of layers in the network
-        layerSizes (list of int): Layer sizes
-        inputSize (int): Size of the input
-        outputSize (int): Size of the output
-        maxLayersize (int): Size of largest layer
-        inputMinimums (list of float): Minimum value for each input
-        inputMaximums (list of float): Maximum value for each input
-        inputMeans (list of floats): Mean value for each input
-        inputRanges (list of floats): Range for each input
-        outputMean (float): Mean value of outputs
-        outputRange (float): Range of output values
-        weights (list): Network weight matrices, where the outer index corresponds to layer number
-        biases (list): Network bias vectors, where the outer index corresponds to layer number
-     """
-    def __init__ (self, filename, normalize=False):
+    """ Abstract class representing a Marabou network from an NNet file
```

Comment:
Remove "Abstract", since this is not an abstract class. There should be an "Attributes:" section that lists attributes of this class (not the ones from parent class). The attributes list should not be separated by where they were initialized. It should follow the same format as that of the original MarabouNetworkNNet docstring. Also, for consistency, the description for f_variables should begin with capital "List" instead of "list".

Reply:
Also, there should be an Args section here, and the docstring for \_\_init\_\_ function should be omitted.

The Args section in the original docstring has a couple of errors. The filename argument doesn't have a type, and the colon should come after the type. So it should look something like:

filename (str): path to ...
normalize (bool): True if ...

It would be great to fix my mistake here.

Reply:
@kjulian3, thanks for the helpful and detailed comments! I am now working on making the changes, and I wanted to ask your advice on something. I am not sure how to treat  attributes (members, not methods) of the class that are not initialized in __init__(): e.g., b_variables in this case. On the one hand, it makes sense to me to include them in the docustring (they were omitted previously). On the other hand, it also makes sense to me to separate them from the other members, if nothing else, as a warning to the developer (after all, documentation should be helpful to us as well as the users -- at least, I believe that it would have been helpful for me if all the attributes had been listed in the docustring when I had started making changes, and it would have also been [mildly] helpful if there had been an indication to where things first "appear"). This is why originally I indicated explicitly that they are initiated elsewhere). Any suggestions/thoughts?

Reply:
It looks like *init*() calls *self.computeNetworkAttributes*(), which call *self.variableRanges*(), which sets the attributes b_variables and f_variables. This happens any time init is called and filename is not empty. If filename is empty, then *self.clearNetwork*() is used instead, so I think b_variables and f_variables should be initialized there to empty lists. 

You could also call clearNetwork at the beginning of init regardless of filename, which would explicitly set all attributes to the default empty/0 values. Then if filename is not empty, the read_nnet and computeNetworkAttributes can be called to update the nominal values.

If finding where a variable is initialized is tricky, we could add "Initialized in self.variablesRanges()" to the attribute's description, for example.

Reply:
Thanks, this sounds good.